File: brightness/python/dpAux.py
#!/usr/bin/python3
#
#    dpAux.py                  (C) 2022-2023, Aurélien Croc (AP²C)
#
#  This program is free software; you can redistribute it and/or modify it under
#  the terms of the GNU General Public License as published by the Free Software
#  Foundation; version 2 of the License.
#
#  This program is distributed in the hope that it will be useful, but WITHOUT
#  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
#  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
#  details.
#  
#  You should have received a copy of the GNU General Public License along with
#  this program; If not, see <http://www.gnu.org/licenses/>.

class DPAux :
    def __init__(self) :
        self.f = None

    def open(self, path='/dev/drm_dp_aux0') :
        self.f = open(path, 'r+b', buffering=0)

    def reopen(self) :
        self.f.close()
        self.open()

    def read(self, addr) :
        self.f.seek(addr, 0)
        return int.from_bytes(self.f.read(1), byteorder="little")

    def write(self, addr, val) :
        self.f.seek(addr, 0)
        return self.f.write((val).to_bytes(1, byteorder='little', signed=False))

    def writeArray(self, array) :
        for (addr, val) in array :
            # Send multiple bytes
            if isinstance(val, list) or isinstance(val, tuple) :
                self.f.seek(addr, 0)
                for i in val :
                    self.f.write((i).to_bytes(1, byteorder='little', signed=False))
            # Send single byte
            else :
                self.write(addr, val)

File: brightness/dpAux.c
/*
 *      tables.c                (C) 2022-2023, Aurélien Croc (AP²C)
 *
 *   This program is free software; you can redistribute it and/or modify it under
 *   the terms of the GNU General Public License as published by the Free Software
 *   Foundation; version 2 of the License.
 * 
 *   This program is distributed in the hope that it will be useful, but WITHOUT
 *   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 *   FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 *   details.
 *   
 *   You should have received a copy of the GNU General Public License along with
 *   this program; If not, see <http://www.gnu.org/licenses/>.
 */
#include <unistd.h>
#include <fcntl.h>
#include "dpAux.h"



aux_t dpAuxOpen(const char *path)
{
    int fd;
   
    if ((fd = open(path, O_RDWR)) == -1)
        return 0;
    return fd;
}

char dpAuxRead(aux_t aux, int addr)
{
    char res;

    lseek(aux, addr, SEEK_SET);
    read(aux, &res, 1);
    return res;
}

void dpAuxWrite(aux_t aux, int addr, char val)
{
    lseek(aux, addr, SEEK_SET);
    write(aux, &val, 1);
}

void dpAuxWrites(aux_t aux, dpWrite_t data[])
{
    for (int i=0; data[i].addr; i++) {
        lseek(aux, data[i].addr, SEEK_SET);
        write(aux, data[i].val, data[i].nVal);
    }
}

void dpAuxClose(aux_t aux)
{
    close(aux);
}



/* vim: set expandtab tabstop=4 shiftwidth=4 smarttab tw=80 cin fenc=utf8 : */

